### PR TITLE
More edge cases for changeCase function and slight behaviour changes.

### PR DESCRIPTION
**Pull request by FSchumacher**
## Description
Addition of more examples to test with the different change case modes.
We now really split on all non alpha numeric characters - even unicode ones.
The input strings for the camel case modes get trimmed before further actions.

A further question is, whether we should trim the input in the other modes as well?
For example "` space in front`" + `capitalize` = "`Space in front`" or "` space in front`"?

## Motivation and Context
The handling of whitespace and other non alphanumeric characters was not well defined.

## How Has This Been Tested?
unit tests are added and run

## Types of changes
- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

File: test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.jmeter.functions;
+
+import static org.junit.Assert.*;
```
Comment (ham1) [inline test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java]: No `*` imports :)

File: test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class TestChangeCaseExamples extends JMeterTestCase {
+
+    @Parameters
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[][] { { " spaces before and after ",
+                new String[] { " SPACES BEFORE AND AFTER ",
```
Comment (ham1) [inline test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java]: One per line or all on one line would make this easier to read, also a comment regarding which mode corresponds to which position would be helpful for future reference (if/when it changes).

File: test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java (hunk 1 above)
Comment (FSchumacher) [reply test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java]: Right, but we don't enforce it yet and there are more in our test code base.

I will have a look at my eclipse settings.

File: test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java (hunk 2 above)
Comment (FSchumacher) [reply test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java]: This would be a perfect showcase for spock, don't you think?

Comment (ham1) [reply test/src/org/apache/jmeter/functions/TestChangeCaseExamples.java]: I agree.

See the PR for an example in `HTTPUtilsSpec` (and smaller examples in a few of the other tests).

File: src/functions/org/apache/jmeter/functions/ChangeCase.java [hunk 3]
```diff
@@ -127,19 +129,14 @@ public String getReferenceKey() {
...
-            String lowerCased = StringUtils.lowerCase(tokens[i]);
-            if(i == 0) {
-                builder.append(isFirstCapitalized ? StringUtils.capitalize(lowerCased):
-                    lowerCased);
-            } else {
-                builder.append(StringUtils.capitalize(lowerCased));
-            }
+    private static String camel(String input, boolean uncapitalizeFirst) {
+        List<String> tokens = Arrays.asList(SPLIT_CHARS.split(input)).stream()
+                .filter(s -> !s.isEmpty()).map(StringUtils::lowerCase).map(StringUtils::capitalize)
```
Comment (ham1) [inline src/functions/org/apache/jmeter/functions/ChangeCase.java]: I generally prefer each item on a new line - I had to read this line twice as I missed the map in the middle on first pass, whereas I think:
```java
.filter(StringUtils::isNotEmpty)
.map(StringUtils::lowerCase)
.map(StringUtils::capitalize)
```
is just that bit easier to read and doesn't get in the way of comprehending the code.

**Comment by pmouawad** [comment]
+1 for trim
Ok for me to merge.

**Comment by FSchumacher** [comment]
@pmouawad do you mean trim for all cases (upper, lower and capitalize)?

**Comment by pmouawad** [comment]
Hi Felix,
It seems there is a consensus to drop CamelCase from function.
I agree also with it as I have never really needed in all my tests camel case nor any member of our team. So I have a doubt  whether it would be useful.
